components/amara/amara.app.mjs

```diff
@@ -0,0 +1,625 @@
+          nextResources,
+          resourceNotFound: true,
+        };
+      }
+
+      const lastResourceIndex =
+        nextResources.findIndex((resource) => JSON.stringify(resource) === lastResourceStr);
+
+      const resourceNotFound = lastResourceIndex < 0;
+
+      nextResources =
+        resourceNotFound
+          ? nextResources
+          : nextResources.slice(0, lastResourceIndex);
+
+      return {
+        resources: resources.concat(nextResources),
+        nextResources,
+        resourceNotFound,
+      };
+    },
+    /**
+     * getResourcesStream always gets the latest resources from the API.
+     * @param {Object} args - all arguments to pass to the `paginateResources` function
+     * @param {Function} args.resouceFn - the name of the resource function to call
+     * @param {Object} args.resourceFnArgs - the arguments object to pass to the resource function
+     * @param {number} [args.offset] - the offset to start at
+     * @param {number} [args.limit=100] - the number of resources to get per page
+     * @param {number} [args.max=300] - the maximum number of resources to get
+     * @param {string} [args.lastResourceStr] - the last resource string in cache
+     * to validate against. This parameter is only passed in from sources.
+     * @returns {Iterable} - Iterable that yields resources,
+     *  the first element is the last resource string in cache
+     */
+    async *getResourcesStream({
+      resourceFn,
+      resourceFnArgs,
+      offset,
+      limit = constants.DEFAULT_PAGE_LIMIT,
+      max = constants.DEFAULT_MAX_ITEMS,
+      lastResourceStr,
+    }) {
+      let resources = [];
+      let lastUrl;
+      let resourceNotFound = true;
+
+      while (true) {
+        const nextResponse = await resourceFn({
+          ...resourceFnArgs,
+          url: lastUrl,
+          params: {
+            ...resourceFnArgs.params,
+            limit,
+            offset,
+          },
+        });
+
+        if (!nextResponse) {
+          throw new Error("No response from the Amara API.");
+        }
+
+        let nextResources = nextResponse.objects;
+
+        ({
+          resources,
+          nextResources,
+          resourceNotFound,
+        } = this.processResources({
+          resources,
+          nextResources,
+          lastResourceStr,
+        }));
+
+        if (nextResponse.meta.next) {
+          lastUrl = nextResponse.meta.next;
+        }
+
+        for (const resource of nextResources) {
+          yield resource;
+        }
+
+        // if resourceNotFound is true, then keeps looping until found
+        // if nextResponse?.meta.next is not null, then there are more resources to fetch
+        const resourceNotFoundAndNexUrl = resourceNotFound && nextResponse?.meta.next;
+
+        // if resources.length < max is the max number of resources to be fetched
+        const keepFetching = resourceNotFoundAndNexUrl && resources.length < max;
```

Comment:
I think changing this method to a generator function allows us to simplify it by removing the call to `this.processResources()`.

```suggestion

      while (true) {
        const nextResponse = await resourceFn({
          ...resourceFnArgs,
          url: lastUrl,
          params: {
            ...resourceFnArgs.params,
            limit,
            offset,
          },
        });

        if (!nextResponse) {
          throw new Error("No response from the Amara API.");
        }

        let nextResources = nextResponse.objects;

        if (nextResponse.meta.next) {
          lastUrl = nextResponse.meta.next;
        }

        for (const resource of nextResources) {
          if (lastResourceStr && JSON.stringify(resource) === lastResourceStr) {
            return;
          }
          yield resource;
        }

        const keepFetching = nextResponse?.meta.next && resources.length < max;
```